Bluetooth: Add additional attributes to DIS

hersp opened the pull request:
Add SW, HW, FW & Serial attributes to DIS
Add support for reading attributes from settings

Signed-off-by: Peter Herager <pehe@oticon.com>

Vudentz commented on subsys/bluetooth/services/dis.c:
```diff
@@ -38,6 +47,33 @@ static struct dis_pnp dis_pnp_id = {
+#define DIS_STORAGE const
+#endif
+
+static DIS_STORAGE u8_t dis_model[DIS_STR_MAX] = CONFIG_BT_GATT_DIS_MODEL;
```
I don't get why we need to declare a variable to store the config options? It seems to be we would be wasting memory as one would have to assign CONFIG_BT_GATT_DIS_STR_MAX with the biggest string size.

hersp replied:
This is only the case when reading from the settings system. Otherwise it is just a const with length defined by the kconfig.

hersp replied:
Can see the the const concept poses some problems during checking, I will work out something else.

hersp replied:
Now using kconfig directly when not using settings system

Vudentz commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -646,6 +646,13 @@ config BT_DEBUG_SDP
+	help
+		This option enables debug support for the Bluetooth
+		Device Information Service(DIS).
+
```
Id split this change in a separate patch.

hersp replied:
I can do that

Vudentz replied:
I spoke to Johan about this change, we agreed that it better to have just one option which enables debug for every service as oppose to have one per service where we may end up bloating the system with debug options. Also the help indentation is wrong, it seems you have copied it from SDP which is also wrong (it should be just 2 spaces.)

hersp replied:
Agree

hersp replied:
Changed

Vudentz commented on subsys/bluetooth/services/Kconfig.dis:
```diff
@@ -7,9 +7,24 @@
+config BT_GATT_DIS_DYNAMIC
+	bool "Enable Dynamic data in Device Information Service"
+	help
+	 Enable Dynamic data in Device Information Service.
```
Im not quite happy about the use of Dynamic term here, it perhaps lead to the impression that it can be changed at runtime or that it is writeable, maybe Options would be better here?

hersp replied:
Copied the term from the device name that can be changed runtime. So no problem changing it to settings.

hersp replied:
Changed

Vudentz commented on subsys/bluetooth/services/dis.c:
```diff
@@ -59,21 +95,145 @@ static ssize_t read_pnp_id(struct bt_conn *conn,
+#endif
+}
+
+BT_SETTINGS_DEFINE(dis, dis_set, NULL, NULL);
```
So this is loading some setting saved on flash but there is no code to store these so I wonder how it gets there in the first place? Furthermore, we do have STR options to set these which gets overwritten if set in flash, Id say we might not need these settings if that can be configured via Kconfig that is probably much easier to setup then have it somewhat stored manually just to be loaded. The general guidance for using the settings subsystem is for values that do change at runtime and need to be persisted, if it is just hardcoded as config option that should not need a setting even if there is a new firmware to be flashed that can just change the config option value.

hersp replied:
In the system I am working on these values can be changed after final compilation as the device could be sold under different brands with each their own versioning scheme. Thus these values will be written in the settings during production and read from setting during boot. so kconfig is not an option.

Vudentz replied:
> In the system I am working on these values can be changed after final compilation as the device could be sold under different brands with each their own versioning scheme. Thus these values will be written in the settings during production and read from setting during boot. so kconfig is not an option.

So the idea is that it always compiled with the same value which always gets overwritten by what is stored in the settings? Im not sure if there is a practical difference in compiling with a different prj.conf though versus fabricating the settings in flash, afaik for doing the latter there is no tool yet, or there is something on the works? Either way with this design we end up having 3 copies of the same string, one for config option, another stored as setting and yet another for storing the either one of the above, so I can see companies actually favoring rebuilding with a different prj.conf rather than fabricating the flash with settings like you are suggesting. Another question about this design is if by trying to fabricate the flash wouldn't this wipe out existing settings? Afaik to do this properly on an existing image one would have to open the flash fs, read all the stored settings (keys, CCC, etc) and then reconstruct the fs, with the config option alone I beleave this would not be necessary because that doesn't change the flash where the settings are stored.

hersp replied:
The solution will not prevent compiling with different prj.conf files and using this way of setting up the system but will just provide an alternative way of controlling this for those who needs this. The default one will be the way it is working today.
I do not know of any tools providing being in the works, but e.g. blue address, IRK and device name will be read during boot. They will have to come from somewhere - don't know how this is done.
BTW. In my system the flash(settings) are generated much later in the flow than the compilation of zephyr as it is part of a larger setup.

Vudentz replied:
Right, though it is still unclear to me how DFU would work with these values stored as settings, or you are saying that we can do partial writes that just overwrites some blocks? Perhaps there should be another Kconfig just to enable the settings loading, that way we only need to enable the extra storage if the settings loading is enabled otherwise you can assign the config options directly as user_data of the attributes.

hersp replied:
I changed it so now, when settings are not used it will work as previously using the kconfig.

My knowledge about the DFU is limited, as it is not used in the system I work on.

hersp committed:
Bluetooth: Add debug kconfig for bluetooth services

Add kconfig to enable debug output for bluetoth services

Signed-off-by: Peter Herager <pehe@oticon.com>

jhedberg commented on subsys/bluetooth/services/dis.c:
```diff
@@ -38,6 +47,48 @@ static struct dis_pnp dis_pnp_id = {
 };
 #endif
 
+#ifdef CONFIG_BT_GATT_DIS_SETTINGS
```
minor: we try to be consistent and use `if defined(CONFIG_BT_...)` throughout the Bluetooth subsystem. It doesn't make much difference for a single condition, but for multiple ones it has the benefit that you can chain them with `||` or `&&` (something you can't do with `#ifdef CONFIG_BT...`).

hersp replied:
OK - Changed - Did not know.

jhedberg commented on subsys/bluetooth/services/Kconfig.dis:
```diff
@@ -7,9 +7,24 @@
 
 menuconfig BT_GATT_DIS
 	bool "Enable GATT Device Information service"
+	default n
```
`n` is the implicit default, so an explicit `default n` should never be used (sorry for not noticing this earlier and causing these extra review rounds)

hersp replied:
Done

jhedberg reviewed (changes requested):
One more minor thing. Also, please be consistent with capital `B` for `Bluetooth` in the commit message. We just discussed this on #bluetooth on slack today.

hersp committed:
Bluetooth: Add additional attributes to DIS

Add SW, HW, FW & serial attributes to DIS
Add support for reading attributes from settings during boot

Signed-off-by: Peter Herager <pehe@oticon.com>

Vudentz reviewed (approved):
LGTM

carlescufi commented on subsys/bluetooth/services/Kconfig.dis:
```diff
@@ -85,4 +99,47 @@ config BT_GATT_DIS_PNP_VER
+	help
+	 Enable Serial Number characteristic in Device Information Service.
+
+config BT_GATT_DIS_FIRMWARE_REVISION
```
Can we make those shorter by using `FW`, `HW` and `SW` instead of `FIRMWARE`, `HARDWARE` and `SOFTWARE`?

hersp replied:
Can do that

carlescufi commented on subsys/bluetooth/services/dis.c:
```diff
@@ -38,6 +47,54 @@ static struct dis_pnp dis_pnp_id = {
+#else /* CONFIG_BT_GATT_DIS_SETTINGS */
+
+#define BT_GATT_DIS_MODEL_REF			\
+		CONFIG_BT_GATT_DIS_MODEL
```
I wonder if we should cast to `const` to make sure those strings are stored in ROM and not RAM

hersp replied:
Not sure what to do here. You cannot just put the string into a const variable as there is no guarantee that you can reference a const variable in ROM as a pointer(depending on compiler/processor).

carlescufi reviewed (commented):
Thanks for this welcome change! Added a couple of minor comments

lemrey commented on subsys/bluetooth/services/dis.c:
```diff
@@ -38,6 +47,46 @@ static struct dis_pnp dis_pnp_id = {
 
+#if defined(CONFIG_BT_GATT_DIS_SETTINGS)
+#define DIS_STR_MAX
+static u8_t dis_model[CONFIG_BT_GATT_DIS_STR_MAX] = CONFIG_BT_GATT_DIS_MODEL;
```
All these can be `const`.

hersp replied:
It is not consts and also you cannot take the address of a const in all systems.

lemrey commented on subsys/bluetooth/services/dis.c:
```diff
@@ -38,6 +47,46 @@ static struct dis_pnp dis_pnp_id = {
 #endif
 
+#if defined(CONFIG_BT_GATT_DIS_SETTINGS)
+#define DIS_STR_MAX
```
Where is this used?

hersp replied:
Oups! Remains from previous version. Will be removed

lemrey reviewed (approved):
LGTM, left two minor comments.

joerchan commented on subsys/bluetooth/services/dis.c:
```diff
@@ -59,15 +107,133 @@ static ssize_t read_pnp_id(struct bt_conn *conn,
 );
+
+#if defined(CONFIG_BT_SETTINGS) && defined(CONFIG_BT_GATT_DIS_SETTINGS)
+static int dis_set(int argc, char **argv, size_t len_rd, settings_read_cb read,
```
rename read argument to read_cb, to avoid confusion?

hersp replied:
Done

joerchan commented on samples/bluetooth/peripheral_dis/src/main.c:
```diff
@@ -44,6 +45,69 @@ static struct bt_conn_cb conn_callbacks = {
+				0xa, 0xb, 0xc, 0xd, 0xe, 0xf };
+
+	memcpy(addr.a.val, addr_val, sizeof(addr_val));
+	settings_runtime_set("bt/id", &addr, sizeof(bt_addr_le_t));
```
Any good reason for doing this? This should be removed in my opinion.

hersp replied:
As settings are enabled now this will provide the a backend for reading.

joerchan replied:
But why do you need to store an IRK and Identity Address in settings? That part I don't understand.

jhedberg replied:
This is also unclear to me. Furthermore, we don't support storing/restoring of a public address - only static random address, so this would fail later either way.

hersp replied:
Agreed. I will remove the non DIS parts.

joerchan reviewed (changes requested):
Looks good. Some minor comments.

jhedberg commented on samples/bluetooth/peripheral_dis/src/main.c:
```diff
@@ -44,6 +45,57 @@ static struct bt_conn_cb conn_callbacks = {
 	.disconnected = disconnected,
 };
 
+
```
Redundant empty line (there should never be a need to have two consecutive ones)

hersp replied:
OK

jhedberg commented on subsys/bluetooth/services/dis.c:
```diff
@@ -59,15 +107,134 @@ static ssize_t read_pnp_id(struct bt_conn *conn,
+BT_SETTINGS_DEFINE(dis, dis_set, NULL, NULL);
+#endif /* CONFIG_BT_GATT_DIS_SETTINGS && CONFIG_BT_SETTINGS*/
+
+
```
Two unnecessary empty lines at end of file, please remove

hersp replied:
OK

jhedberg reviewed (changes requested):
This looks generally good to me, with the exception of some redundant whitespace (empty lines). Could you fix those, and then we can merge this.

hersp commented on subsys/bluetooth/services/dis.c:
```diff
@@ -38,6 +47,45 @@ static struct dis_pnp dis_pnp_id = {
+
+#endif /* CONFIG_BT_GATT_DIS_SETTINGS */
+
+
```
OK

jhedberg replied:
Redundant empty line

hersp commented on samples/bluetooth/peripheral_dis/prj.conf:
```diff
@@ -1,5 +1,28 @@
 CONFIG_BT=y
 CONFIG_BT_PERIPHERAL=y
+
+
```
OK

jhedberg replied:
No need for two empty lines, just use one

hersp commented on samples/bluetooth/peripheral_dis/prj.conf:
```diff
@@ -1,5 +1,28 @@
+
+CONFIG_BT_GATT_DIS_SETTINGS=y
+CONFIG_BT_GATT_DIS_STR_MAX=21
+
```
OK

jhedberg replied:
Extra empty line at the end of the file, please remove
